### PR TITLE
Restore logging of test configuration for pilot e2e tests

### PR DESCRIPTION
**Pull request by nmittler**
This was removed by #3502

To cleanly do this, I've separated out the configuration from the
test runtime environment.  I've renamed `Infra` to `Environment` to
make this clear.

Because fields are now split across `Config` and `Environment`, I've
also create a new `TemplateData` struct that merges all the common
template parameters into a single struct. This way the template files
do not have to change. The separation of template params from the
environment should also help a bit with maintenance going forward.

### CODE REVIEW

**Comment by mattdelco** [comment]
Is the --logtostderr also worthwhile to restore?  Offhand I thought misc tests were passing this (and circle still is in a few places) but since everybody was I think it was migrating towards this common location in the makefile.

**Comment by nmittler** [comment]
@mattdelco 

> Is the --logtostderr also worthwhile to restore? Offhand I thought misc tests were passing this (and circle still is in a few places) but since everybody was I think it was migrating towards this common location in the makefile.

The CircleCI `TESTOPTS` already includes `logtostderr`.  I've updated prow to include it as well.

**Comment by ldemailly** [comment]
linter not happy ^

**Comment by nmittler** [comment]
@ldemailly should be good to go now ... mind re-approving?

**Comment by nmittler** [comment]
@ldemailly any ideas why the tests are failing? Has something changed with the test environment?

**Comment by ldemailly** [comment]
@nmittler  you need to urgently fix https://github.com/istio/istio/issues/3499
in this PR for instance

**Comment by mattdelco** [comment]
One aspect that probably did change on you was that KUEBCONFIG wasn't being set and that was fixed about a half day ago (the earlier successful test run for this PR logged "Env variable KUBECONFIG not set. Skipping tests").

**Comment by ijsnellf** [comment]
Let's keep this PR focused. #3499 and #3064 deserve their own PRs.

**Comment by nmittler** [comment]
@ldemailly PTAL ... I've broadened the scope of this PR to include several fixes to the pilot e2e tests

**Comment by ldemailly** [comment]
there is an interesting failure in the log (https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/3459/istio-pilot-e2e/3042/), reminds me of the 423 issue @kyessenov was chasing down a while ago:
///
02-962d-b4e2-e2662222e806\n2018/02/15 01:48:52 [32 body] User-Agent=Go-http-client/1.1\n2018/02/15 01:48:52 [32 body] Connection=Upgrade\n2018/02/15 01:48:52 [32 body] Sec-Websocket-Key=L4aqGigQexU8i5zKm6mBbA==\n2018/02/15 01:48:52 [32 body] Sec-Websocket-Version=13\n2018/02/15 01:48:52 [32 body] Testwebsocket=enabled\n2018/02/15 01:48:52 [32 body] Upgrade=websocket\n2018/02/15 01:48:52 [32 body] X-Forwarded-Proto=http\n2018/02/15 01:48:52 [7 body] Upgrade=websocket\n2018/02/15 01:48:52 [7 body] Testwebsocket=enabled\n2018/02/15 01:48:52 [7 body] Content-Length=0\n2018/02/15 01:48:52 [7 body] X-Forwarded-Proto=http\n2018/02/15 01:48:52 [7 body] X-Request-Id=4ea9963c-adf7-93b6-82d9-a285cf01860e\n2018/02/15 01:48:52 [7 body] Connection=Upgrade\n2018/02/15 01:48:52 [7 body] Sec-Websocket-Key=JwebAFhNKMDR1OjaDk3H1A==\n2018/02/15 01:48:52 [7 body] Sec-Websocket-Version=13\n2018/02/15 01:48:52 [7 body] Hostname=c-v1-67956fd759-q79kg\n2018/02/15 01:48:52 [7 body] HelloWorld\n2018/02/15 01:48:52 [32 body] Hostname=c-v1-67956fd759-q79kg\n2018/02/15 01:48:52 [32 body] HelloWorld\n2018/02/15 01:48:52 [12 body] ServiceVersion=v1\n2018/02/15 01:48:52 [12 body] ServicePort=80\n2018/02/15 01:48:52 [12 body] Method=GET\n2018/02/15 01:48:52 [12 body] URL=/a\n2018/02/15 01:48:52 [12 body] Proto=HTTP/1.1\n2018/02/15 01:48:52 [12 body] RemoteAddr=127.0.0.1:40564\n2018/02/15 01:48:52 [12 body] Host=c\n2018/02/15 01:48:52 [12 body] Content-Length=0\n2018/02/15 01:48:52 [12 body] X-Request-Id=9c1e9cb8-1b52-9d96-8224-f9928408578a\n2018/02/15 01:48:52 [12 body] Sec-Websocket-Key=Otk7WVXRD+FNlsQWauXgNQ==\n2018/02/15 01:48:52 [12 body] Testwebsocket=enabled\n2018/02/15 01:48:52 [12 body] Upgrade=websocket\n2018/02/15 01:48:52 [12 body] Sec-Websocket-Version=13\n2018/02/15 01:48:52 [12 body] X-Forwarded-Proto=http\n2018/02/15 01:48:52 [12 body] User-Agent=Go-http-client/1.1\n2018/02/15 01:48:52 [12 body] Connection=Upgrade\n2018/02/15 01:48:52 [12 body] Hostname=c-v1-67956fd759-q79kg\n2018/02/15 01:48:52 [12 body] HelloWorld\n2018/02/15 01:48:52 [11 body] ServiceVersion=v1\n2018/02/15 01:48:52 [11 body] ServicePort=80\n2018/02/15 01:48:52 [70 body] ServiceVersion=v1\n2018/02/15 01:48:52 [69 body] ServiceVersion=v1\n2018/02/15 01:48:52 [58 body] ServiceVersion=v1\n2018/02/15 01:48:52 [11 body] Method=GET\n2018/02/15 01:48:52 [11 body] URL=/a\n2018/02/15 01:48:52 [11 body] Proto=HTTP/1.1\n2018/02/15 01:48:52 [11 body] RemoteAddr=127.0.0.1:40566\n2018/02/15 01:48:52 [11 body] Host=c\n2018/02/15 01:48:52 [11 body] Connection=Upgrade\n2018/02/15 01:48:52 [11 body] Sec-Websocket-Key=mP5Bf9ZuH4Way+Bq4qsdhA==\n2018/02/15 01:48:52 [70 body] ServicePort=80\n2018/02/15 01:48:52 [69 body] ServicePort=80\n2018/02/15 01:48:52 [58 body] ServicePort=80\n2018/02/15 01:48:52 [11 body] Upgrade=websocket\n2018/
...
that or just some botched logging with escaped \n that shouldn't be

File: tests/e2e/tests/pilot/access_test.go [hunk 1]
```diff
@@ -79,24 +79,24 @@ func (a *accessLogs) check(infra *tutil.Infra) error {
 
 				pod := infra.Apps[app][0]
 				container := inject.ProxyContainerName
-				ns := infra.Namespace
+				ns := infra.Config.Namespace
```
Comment (ldemailly) [inline tests/e2e/tests/pilot/access_test.go]: can you comment/explain why there is an extra object layer now/what does it buy/help with ?

File: tests/e2e/tests/pilot/pilot_test.go [hunk 2]
```diff
@@ -75,93 +97,100 @@ func init() {
...
-
-	if config.Tag == "" {
-		t.Skip("TAG not specified. Skipping integration tests")
-	}
-
-	// TODO(nmittler): Restore this tutil.Tlog("Deploying infrastructure", spew.Sdump(config))
-	if config.Err = config.Setup(); config.Err != nil {
-		t.Fatal(config.Err)
+func setup(infra *tutil.Infra, t *testing.T) {
+	tutil.Tlog("Deploying infrastructure", spew.Sdump(infra.Config))
```
Comment (ldemailly) [inline tests/e2e/tests/pilot/pilot_test.go]: there is probably a subset of the previous object that can be printed without needing a new object no ?
(the diff is quite big just to reduce log flood - and yet the tests do look floody still)

**Comment by nmittler** [comment]
@ldemailly I've split out the testopts change to #3516

**Comment by nmittler** [comment]
@ldemailly PTAL.  This now only addresses the restoring of the test configuration log.  See the updated description for an overview.

**Comment by nmittler** [comment]
@ldemailly the test likely won't pass until #3663 lands

**Comment by ldemailly** [comment]
this still seems to be a huge diff just to change 1 logging ?
why the tutil.Infra -> tutil.Environment mass rename ?

**Comment by nmittler** [comment]
@ldemailly 

> why the tutil.Infra -> tutil.Environment mass rename

Agreed, there are a lot of changes, but it's mostly renames. See the updated description for an overview.

The rename was effectively due the introduction of the new `Config` type. The term `Environment` seems a fairly common term to refer to the run-time environment of a testing framework, and I wanted to clearly disambiguate it from the configuration for that environment.   The rename also gets us closer to the terminology used by the "common" integration testing environment.

**Comment by nmittler** [comment]
@ldemailly gentle ping ... would like to get this in soonish if possible

**Comment by ldemailly** [comment]
I asked Zach to have a glance

As you will own this code I suppose it is fine you do whichever you feel comfortable but I wanted a second opinion

**Comment by nmittler** [comment]
@ZackButcher I had to rebase ... mind re-approving?

**Pull request merged**